src/main/scala/com/github/jberggg/router/RouterService.scala

```diff
@@ -0,0 +1,70 @@
+package com.github.jberggg.router
+
+import cats.syntax.all._
+import cats.effect.syntax.all._
+import cats.effect.kernel.{Async, Resource}
+import cats.effect.std.Dispatcher
+import org.http4s.Uri
+import org.http4s.Uri.Path
+import fs2.concurrent.SignallingRef
+import org.scalajs.dom.window
+import scala.scalajs.js
+import org.scalajs.dom.PopStateEvent
+import Domain._
+
+object RouterService {
+  
+    def createSignalAndResources[F[_] : Async ](initialState: HistoryApiState): Resource[F,SignallingRef[F,Tuple2[Path,HistoryApiState]]] = for {
+        s <- Resource.eval(createPathSignal[F](initialState))
+        _ <- createHistoryApiResources[F](s)
+    } yield s
+
+    def createHistoryApiResources[F[_] : Async ](signal: SignallingRef[F,Tuple2[Path,HistoryApiState]]): Resource[F,Unit] = for {
+        d <- Dispatcher[F]
+        _ <- createHistoryPopEventListener[F](d,signal)
+        _ <- createHistoryStatePusher[F](signal)
+    } yield ()
+
+    def createPathSignal[F[_] : Async ](initialState: HistoryApiState): F[SignallingRef[F,Tuple2[Path,HistoryApiState]]] = 
+        Async[F]
+        .delay( window.location.href )
+        .map( Uri.unsafeFromString )
+        .map( u => Tuple2(u.path, initialState) )
+        .flatTap{ case (p,s) => Async[F].delay(window.history.replaceState(s,"",p.toString)) } // set initial state in History API
+        .flatMap{ SignallingRef.apply[F,Tuple2[Path,HistoryApiState]] }
+
+    def createHistoryStatePusher[ F[_] : Async ](s: SignallingRef[F,Tuple2[Path,HistoryApiState]]): Resource[F,Unit] = 
+        s.discrete
+        .evalMap{ case (p,st) => Async[F].delay( window.history.pushState( st, "", p.toString ) ) }
```

Comment:
I wonder if there is some subtle trickyness here. I understand that the History API supports relative paths, so when you push `p` it navigates relative to your current location. Because you are also storing `p` in the state, you will may get access to it again in the future. But if your location is different at that time, then when you attempt to resolve it relatively, you will actually end up with a different location.

Reply:
If I'm not totally wrong now the location is only relative (to the host) at the time of pushing the state. So when you navigate back or if the user selects a history item the browser will set the URL that is stored in the history (item) and will send the state with the popstate-event. So you will get the correct URL. What I'm wondering now is however how do you receive the state when the user loads the history item from the history e.g. the page was opened yesterday and closed since then but now I want to open it again and select it in my browser history.

Reply:
Yes I believe you are right—when navigating back in history, the location will be correct, but the `Path` object from the state will still be relative.

> What I'm wondering now is however how do you receive the state when the user loads the history item from the history e.g. the page was opened yesterday and closed since then but now I want to open it again and select it in my browser history.

I think this is why it is important that the state is serializable, I believe the MDN docs mentioned that it can get saved into persistent memory.

Reply:
yes, I remember the same with the persistent memory. I will try to find out how the gets populated in the case a history item gets clicked and if that would work with this implementation.

Reply:
As far as I understand it the `session history` (which you can manipulate with the History API) and the "regular" History which you can look at and click items to go back to this page are two different things. I didn't do a deep dive as I lack the time but from studying the MDN Docs and HTML Standard this would make sense.

The [History API site](https://developer.mozilla.org/en-US/docs/Web/API/History) states:
> The History interface allows manipulation of the **browser session history, that is the pages visited in the tab or frame that the current page is loaded in.**

The responsible [HTML Standard section](https://html.spec.whatwg.org/multipage/history.html#the-session-history-of-browsing-contexts):
>The History interface allows manipulation of the browser session history, that is the pages visited in the tab or frame that the current page is loaded in.

and

>Pages can add serialized state to the session history. These are then deserialized and returned to the script when the user (or script) goes back in the history, thus enabling authors to use the "navigation" metaphor even in one-page applications.

So it's not possible to retrieve a previous stored state when clicking an item of the "regular" browser history. But I guess when your browser crashes and it asks you whether you want to restore your session I guess the page navigation still works with state.

Knowing (or think to know) all that I'm not sure that I will use this features in my personal projects anyways. The only use case I see at the moment is to restore customized overview pages e.g. a list of items filtered by their type and maybe sorted in a custom way. That would allow the user to open a detail page of a specific item and then jump back without losing the custom sorting / filtering. But then again you could also encode that into the URL. Another use case may be forms that span multiple pages.

Reply:
> Knowing (or think to know) all that I'm not sure that I will use this features in my personal projects anyways.

Yes, I had the same feeling. I need to check some of the other routers to see if they do anything interesting with it. Otherwise, I think this is one of the reasons I thought separating the `History` and `Router` DSLs could be a good strategy: the `History` DSL can expose the state for power users, but the `Router` DSL can focus on simplicity and not expose it.

Reply:
Yes agree, then such a split makes sense to me as well 👍🏽